AppiumFuncTests/AppiumDriver/AppiumFuncTests.swift

```diff
@@ -193,4 +193,36 @@ class AppiumFuncTests: XCTestCase {
         guard let countAfter = eventsAfter?.events.count else { return }
         XCTAssertTrue(countAfter > countBefore)
     }
+
+    func testCanTypeTextOntoTextField() {
+        try! driver.findElement(by: .name, with: "TextFields").click()
+        let textFld = try! driver.findElement(by: .name, with: "Rounded")
+        textFld.click()
+        textFld.sendKeys(with: "Send Keys Test")
+        // TODO: refactor this assertion once https://appium.io/docs/en/commands/element/attributes/text/ is implemented
+        XCTAssertTrue(try! driver.getPageSource().get().contains("Send Keys Test"))
+    }
+
+    func testCantTypeTextOntoInvalidElements() {
+        /** Something odd happens here: if driver tries to type text onto a button, the first keystroke actually clicks the target button,
+         changing screens and throwing an "invalid element state" error. Not sure if this is expected behaviour or a WebDriverAgent bug?
```

Comment:
mm, let me check the WDA code.
Probably newer WDA simply calls type method in XCTest framework, but older one tap the target element once to ensure keyboard appears and ready to type texts as a workaround for older XCTest framework.